### PR TITLE
[Test][Multi-user] Set up multi user e2e test 

### PR DESCRIPTION
**Pull request by chensun**
Part of #3288
* Deploy test cluster using Kubeflow full-fledge deployment.
* Run full multi-user end to end test using SDK.
* Add instruction on how to run multiuser test locally.

/cc @IronPan @Bobgy

### CODE REVIEW

File: test/deploy-multiuser-kfp.sh [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+  set +e # the following clean up commands shouldn't exit on error
+  set +x # no need for command history
+
+  echo "Status of pods before clean up:"
+  kubectl get pods --all-namespaces
+
+  echo "Clean up service accounts..."
+  for sa_suffix in 'admin' 'user' 'vm';
+  do
+      gcloud iam service-accounts delete ${TEST_CLUSTER}-${sa_suffix}@${PROJECT}.iam.gserviceaccount.com --quiet
```
[the remark this answers is not in the document]
Comment (chensun) [reply test/deploy-multiuser-kfp.sh:27]: Thanks! I want to discuss this topic as well. 
For the short term, I was hoping that the rate of cleanup failures may be low (depending on how frequently we want to run the test, for instance, if once per day), and when we do hit the quota issue, hopefully not too often, oncall may fix this manually. (not an ideal solution, I agree)
For the long term solution, I wonder if we should instead adopt this [script](https://github.com/kubeflow/testing/blob/master/py/kubeflow/testing/cleanup_ci.py), which already covers more cleanups than our current tool. Be aware that we're currently leaking resources already even without multi-user tests. I previously did a manual cleanup, and opened an internal [bug](b/149244649) for tracking. This is what brought my attention to the above script.

File: test/deploy-multiuser-kfp.sh [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+download-kustomize
+
+deploy-kubeflow-cli-env $TEST_CLUSTER
+
+mkdir -p $KF_DIR
+pushd $KF_DIR
+
+kfctl build -V -f ${CONFIG_URI}
+
+# Deploy test images
```
[the remark this answers is not in the document]
Comment (Bobgy) [reply test/deploy-multiuser-kfp.sh:123]: lol, that's true!

File: test/multiuser-tests.sh [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+  retry gcloud projects add-iam-policy-binding $PROJECT \
+    --member="serviceAccount:$USER_GSA@$PROJECT.iam.gserviceaccount.com" \
+    --role="roles/editor"
+fi
+
+# Create test profiles in prepare for test scenarios
+kubectl create -f ${DIR}/multi-user-test/profile.yaml
+
+KFP_HOST="https://${TEST_CLUSTER}.endpoints.${PROJECT}.cloud.goog/pipeline"
+CLIENT_ID=${CLIENT_ID:-$(gsutil cat gs://ml-pipeline-test-keys/oauth_client_id)}
```
[the remark this answers is not in the document]
Comment (Bobgy) [reply test/multiuser-tests.sh:118]: How much effort is it? I'd prefer credentials regenerated.
They can run pipelines in our test clusters, and modify other GCP settings/resources (the test pipelines have a lot of privilege).

File: manifests/kustomize/wi-utils.sh [hunk 4]
```diff
@@ -64,12 +65,12 @@ function verify_workload_identity_binding {
   for i in $(seq 1 ${max_attempts})
   do
     workload_identity_is_ready=true
-    kubectl run test-$RANDOM --rm -i --restart=Never \
+    kubectl run test-$RANDOM --rm -i \
```
[the remark this answers is not in the document]
Comment (Bobgy) [reply manifests/kustomize/wi-utils.sh:67]: Thanks for the explanation, would be great putting these as a comment next to it.

File: test/deploy-multiuser-kfp.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (Bobgy) [reply test/deploy-multiuser-kfp.sh:27]: That's a good suggestion! Makes sense to me.
Can you log a github issue?

File: test/test-prep.sh [hunk 5]
```diff
@@ -16,7 +16,7 @@
 
 set -x
 
-ZONE=${ZONE:-us-east1-b}
+ZONE=${ZONE:-us-east1-c}
```
[the remark this answers is not in the document]
Comment (Bobgy) [reply test/test-prep.sh:19]: Got it. Thanks!

File: test/multiuser-tests.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (chensun) [reply test/multiuser-tests.sh:118]: Regenerated two clients' secrets and three users' tokens: https://screenshot.googleplex.com/0GbuBxsYqXK

File: manifests/kustomize/wi-utils.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (chensun) [reply manifests/kustomize/wi-utils.sh:70]: Added comments

File: test/deploy-multiuser-kfp.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (chensun) [reply test/deploy-multiuser-kfp.sh:27]: Opened #3691

[the remark this answers is not in the document]
Comment (Bobgy) [reply test/deploy-multiuser-kfp.sh]: Thanks!

File: test/multiuser-tests.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (Bobgy) [reply test/multiuser-tests.sh:118]: Great, thanks!

**Comment by Bobgy** [comment]
/lgtm
/approve
/hold

Can we configure this as an optional presubmit test now?

**Comment by chensun** [comment]
> /lgtm
> /approve
> /hold
> 
> Can we configure this as an optional presubmit test now?

By optional, do you mean we run it all the time but the result, if it's a failure, doesn't block merging? or do you mean the test is on and off randomly? First option is more favorable, IMO.
Does our current infra support these? Thanks!

**Comment by Bobgy** [comment]
@chensun You can find test in prow's PR history: https://prow.k8s.io/pr-history/?org=kubeflow&repo=pipelines&pr=3673.

I'm not sure why it doesn't show up in status, probably something was misconfigured in the test config that made it not reported.

**Comment by Bobgy** [comment]
@chensun didn't see you question

> By optional, do you mean we run it all the time but the result, if it's a failure, doesn't block merging? or do you mean the test is on and off randomly? First option is more favorable, IMO.
Does our current infra support these? Thanks!

I meant it only runs when we request it.

**Pull request closed without merging**